### ISSUE TITLE
Service_Generic declaration returns "undefined" error for HTTPS virtual server declaration

### ISSUE THREAD

**Issue by fustilarian01** [F5Networks/f5-appsvcs-extension#304]
### Environment
 * Application Services Version: 3.20.0.3
 * BIG-IP Version: 15.1.0.2

### Summary
Posting a simple virtual server declaration using the Service_Generic class returns the error "Cannot read property 'undefined' of undefined", however the same declaration using the Service_HTTPS will post successfully.

### Steps To Reproduce
1. Submit the following declaration:
```json
{
    "class": "ADC",
    "schemaVersion": "3.20.0",
    "id": "Initial_deployment",
    "Common": {
        "class": "Tenant",
        "label": "commonTenant",
        "remark": "Declaration Common tenant",
        "enable": true,
        "Shared": {
            "class": "Application",
            "template": "shared",
            "main_https": {
                "class": "Service_Generic",
                "profileHTTP": {
                    "bigip": "/Common/http"
                },
                "profileTCP": {
                    "bigip": "/Common/tcp"
                },
                "virtualAddresses": [
                    "10.0.0.0/30"
                ],
                "virtualPort": 443,
                "snat": "auto",
                "translateClientPort": true,
                "translateServerAddress": true,
                "serverTLS": {"bigip": "/Common/clientssl"},
                "clientTLS": {"bigip": "/Common/serverssl"}
            }
        }
    }
}
```

2. Observe the following error response:
```json
{
    "message": "Cannot read property 'undefined' of undefined",
    "host": "localhost",
    "tenant": "Common",
    "code": 422
},
{
    "message": "Cannot read property 'undefined' of undefined",
    "host": "localhost",
    "tenant": "Common",
    "code": 422
}
```

### Expected Behavior
The virtual server declaration would post successfully

### Actual Behavior
I get an error without any information letting me know what/where the issues is. The same declaration will work if I use the Service_HTTPS class.

**Comment by dstokesf5** [comment]
I tried your declaration on 3.20.0, and it was successful:
```json
{
    "results": [
        {
            "code": 200,
            "message": "success",
            "lineCount": 21,
            "host": "localhost",
            "tenant": "Common",
            "runTime": 5194
        },
        {
            "code": 200,
            "message": "no change",
            "host": "localhost",
            "tenant": "Common",
            "runTime": 2889
        }
    ],
    "declaration": {
        "class": "ADC",
        "schemaVersion": "3.20.0",
        "id": "Initial_deployment",
        "Common": {
            "class": "Tenant",
            "label": "commonTenant",
            "remark": "Declaration Common tenant",
            "enable": true,
            "Shared": {
                "class": "Application",
                "template": "shared",
                "main_https": {
                    "class": "Service_HTTPS",
                    "profileHTTP": {
                        "bigip": "/Common/http"
                    },
                    "profileTCP": {
                        "bigip": "/Common/tcp"
                    },
                    "virtualAddresses": [
                        "10.0.0.0/30"
                    ],
                    "virtualPort": 443,
                    "snat": "auto",
                    "translateClientPort": true,
                    "translateServerAddress": true,
                    "serverTLS": {
                        "bigip": "/Common/clientssl"
                    },
                    "clientTLS": {
                        "bigip": "/Common/serverssl"
                    }
                }
            }
        },
        "updateMode": "selective",
        "controls": {
            "archiveTimestamp": "2020-07-29T16:14:05.294Z"
        }
    }
}
```

Are there additional steps required to reproduce this issue?

**Comment by fustilarian01** [comment]
Nope

**Comment by dstokesf5** [comment]
I am still unable to reproduce this issue with AS3 3.20.0-3, BIG-IP version 15.1.0.5, and the provided declaration. I will need additional information in order to resolve the issue you are seeing. Do you see anything in /var/log/restnoded/restnoded.log on the BIG-IP when you see this failure?

**Comment by m-kratochvil** [comment]
Hi,
I also see this error, when customer tries to configure L4 virtual server with `clientTLS` pool, not realizing this is not supported.
There is an error on the Big-IP in `/var/log/restnonded/restnonded.log` - see below.

Application Services Version: 3.22.0-1
BIG-IP Version: 15.1.2.1

Here is an example declaration that generates the condition:
```
{
    "action": "deploy",
    "class": "AS3",
    "declaration": {
        "class": "ADC",
        "id": "urn:uuid:748894e5-a062-410e-b503-7cd9f3d505f2",
        "label": "F5 BigIP",
        "tenant_partition": {
            "class": "Tenant",
            "defaultRouteDomain": 2763,
            "label": "tenant_1234",
            "lb_test": {
                "class": "Application",
                "label": "384140bd-5c71-4910-b836-567e70dc9230",
                "listener_test": {
                    "class": "Service_L4",
                    "clientTLS": "tls_pool_12345",
                    "iRules": [],
                    "label": "tin_list",
                    "persistenceMethods": [
                        "source-address"
                    ],
                    "policyEndpoint": [],
                    "pool": "pool_12345",
                    "profileL4": {
                        "bigip": "/Common/cc_fastL4_profile"
                    },
                    "snat": "self",
                    "virtualAddresses": [
                        "10.20.30.14/32"
                    ],
                    "virtualPort": 443
                },
                "pool_12345": {
                    "class": "Pool",
                    "label": "tin_pool",
                    "loadBalancingMode": "least-connections-member",
                    "members": [],
                    "remark": "tin_pool"
                },
                "template": "generic",
                "tls_pool_12345": {
                    "class": "TLS_Client"
                }
            }
        },
        "schemaVersion": "3.19.0",
        "updateMode": "selective"
    },
    "historyLimit": 2,
    "logLevel": "warning",
    "persist": false,
    "trace": false
}
```

While I would expect this to fail, there is only one error message - in the `restnoded.log` file, and the message is non descriptive:
```
severe: [appsvcs] {"message":"TypeError: Cannot read property 'undefined' of undefined","level":"error"}
```

Questions:
1. How is this error generated?
2. Could there possibly be a better descriptive error message?

Thanks!

**Comment by dstokesf5** [comment]
@m-kratochvil, the error is coming from the clientTLS property of `tenant_partition/lb_test/listener_test`. It looks like there is a bug when using Service_L4 and specifying a clientTLS with a relative path. If you specify the full path (`/tenant_partition/lb_test/tls_pool_12345`) you get past the error message.

I have added AUTOTOOL-2367 to our internal product backlog to fix up the "TypeError: Cannot read property 'undefined' of undefined" error message.

**Comment by m-kratochvil** [comment]
Thank you for the response, and for committing to fix the error message.

If I understand it right, specifying `ClientTLS` property with class `ServiceL4` will never yield successful declaration anyway, as it's not a supported combination by the Big-IP..

**Comment by dstokesf5** [comment]
Correct, clientTLS is not supported on the Service_L4 class because it is not a combination that is supported by the BIG-IP.

**Comment by dstokesf5** [comment]
This issue was resolved as part of the 3.28.0 release.
